packages/json-api/src/Registry.ts

```diff
@@ -0,0 +1,92 @@
+import { JSONValue, JSONAPIDocument, JSONAPIResource } from "./jsonApiTypes";
+import {
+  IncludesMap,
+  MapObject,
+  Registry,
+  Resource,
+  ResourceParseOptions,
+  ResourceSpec
+} from "./types";
+import { identity, kebabCaseDeep, camelCaseDeep, TransformFunc } from "./utils";
+import ResourceImpl from "./Resource";
+
+export type RegistryCreateOptions = {
+  keyTransform?: "kebab";
+};
+
+export default class RegistryImpl implements Registry {
+  private options?: RegistryCreateOptions;
+
+  private resources: MapObject<Resource>;
+
+  public keyTransformFunc: TransformFunc;
+
+  public keyParseFunc: TransformFunc;
+
+  constructor(options?: RegistryCreateOptions) {
+    this.options = options;
+    this.resources = {};
+
+    this.keyTransformFunc =
+      this.options && this.options.keyTransform === "kebab"
+        ? kebabCaseDeep
+        : identity;
+    this.keyParseFunc =
+      this.options && this.options.keyTransform === "kebab"
+        ? camelCaseDeep
+        : identity;
+  }
+
+  define(type: string, spec: ResourceSpec): Resource {
+    const resource = new ResourceImpl(this, type, spec);
+    this.resources[type] = resource;
```

Comment:
Does resource type overwrite is allowed?

Reply:
Good point, let's throw and error in this case.